Improvements on the generation of silica detritus

jessluo opened the pull request:
Description to come. This is just a placeholder as we work on a fix to improve the silica cycle. Current ESM4.5 runs have a unrealistic build-up of silica at the poles, associated with insufficient sidet generation from phytoplankton mortality.

jessluo committed:
routing phytoplankton mortality Si to si detritus instead of remineralized sio4

jessluo committed:
modifications to handling of silica detritus

jessluo committed:
change defaults and make sure exuloss_sio2 is positive

jessluo committed:
allocate/deallocate jexuloss_sio2, add diagnostics

charliestock commented on generic_tracers/cobalt_reg_diag.F90:
```diff
@@ -798,6 +798,20 @@ subroutine cobalt_reg_diagnostics(diag_list,axes,init_time,phyto,zoo,bact,cobalt
     phyto(SMALL)%id_jexuloss_fe = register_diag_field(package_name, vardesc_temp%name, axes(1:3),&
          init_time, vardesc_temp%longname,vardesc_temp%units, missing_value = missing_value1)
 
+    vardesc_temp = vardesc("jexuloss_sio2_Lg","Large phyto silica loss via exudation",&
```
Maybe we could a comment here saying that "exudation" is the dissolution of si shells in diatoms to sio4 associated with phytoplankton mortality and when respiration exceeds growth.  With this in mind, should this be jdissloss so that it is consisent with the dissolution of sinking sidet?  Also, do we need to define this for diazotrophs if we populate for n = 1,NUM_PHYTO below?

theresa-cordero replied:
@charliestock has this comment been sufficiently addressed?

charliestock commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -1360,7 +1367,14 @@ subroutine user_add_params(param_file)
                    "fraction of P ingestion by large zooplankton to semi-labile dissolved organic phosphorus", &
                    units="none", default=0.3*(0.30-zoo(3)%phi_det))
-    !
+    ! Partitioning of silica detritus production from zooplankton is by default slightly higher than organic matter detritus
```
I think we should make the default equal for now so that we recover answers and override for ESM4.5.  We can eventually switch the defaults to whatever we settle on for ESM4.5.

jessluo replied:
Yes, defaults are set to be 0 for small zooplankton, 0.15 for medium zooplankton, and 0.30 for large zooplankton. it should reproduce answers.

charliestock commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -3923,17 +3937,25 @@ subroutine generic_COBALT_update_from_source(tracer_list,Temp,Salt,rho_dzt,dzt,h
        cobalt%nlg_misc(i,j,k)=phyto(LARGE)%f_n(i,j,k) - phyto(LARGE)%f_n(i,j,k)*phyto(LARGE)%silim(i,j,k)
        cobalt%nmd_misc(i,j,k)=phyto(MEDIUM)%f_n(i,j,k) - phyto(MEDIUM)%f_n(i,j,k)*phyto(MEDIUM)%silim(i,j,k)
+
+	   ! silim present in here twice to first determine the biomass of the diatoms, and secondly to determine the uptake
```
Looking more closely at the code, I think it might be best to say "first to find the fraction of nitrogen uptake attributed to diatoms, and then to scale the Si:N ratio of that uptake.

jessluo replied:
Done.

charliestock commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -4501,7 +4523,7 @@ subroutine generic_COBALT_update_from_source(tracer_list,Temp,Salt,rho_dzt,dzt,h
             phyto(n)%jmortloss_p(i,j,k) = phyto(n)%jmortloss_n(i,j,k)*phyto(n)%q_p_2_n(i,j,k)
             phyto(n)%jmortloss_fe(i,j,k) = phyto(n)%jmortloss_n(i,j,k)*phyto(n)%q_fe_2_n(i,j,k)
-            phyto(n)%jmortloss_sio2(i,j,k) = phyto(n)%jmortloss_n(i,j,k)*phyto(n)%q_si_2_n(i,j,k)
+            phyto(n)%jmortloss_sio2(i,j,k) = phyto(n)%phi_sidiss_mort*phyto(n)%jmortloss_n(i,j,k)*phyto(n)%q_si_2_n(i,j,k)
```
Given the changes above, I wonder if it would make sense to add this to a variable called phyto(n)%jdissloss_sio2(i,j,k) rather than having a separate jmortloss?  That is, phytoplankton mortality results in silica dissolution?

jessluo replied:
Sounds good - changed in the new commit.

charliestock reviewed (commented):
Hi @jessluo,

Thanks for turning this around so quickly.  I think it looks good, but I had a couple of minor questions and suggestions:

1) I think the defaults for the phi_det_si should be equal to the organic variables to reproduce prior answers and make life easier for Theresa.

2) I am wondering whether we should call the combined conversion of silica in the shells of individually sinking diatoms phyto(n)%jdissloss_si rather then separating it into jexuloss_si and jmortloss_si.  The process involved is essentially a dissolution of silica resulting from the net respiration and mortality.  Replacing jexuloss_si and jmortloss_si with a jdiss_si variable containing the sum may be a better way to capture the process and track patterns in remineralization.

3. I had a few very minor suggestions for the comments to consider.

Drop by any time. I am in.

best,
Charlie

jessluo committed:
combined exuloss_sio2 and mortloss_sio2 into a single dissloss_si, cleaned up diagnostics and comments, and fixed source/sink imbalance

jessluo committed:
bug fix

jessluo committed:
bug fix

theresa-cordero reviewed (approved):
Passed tests with NWA12 and NEP10.
I recommend squash merging so these changes are in a single commit.

theresa-cordero commented:
Jessica, can you write a sentence or two for the commit message associated with these changes? Then I will be able to merge it.

jessluo commented:
A more detailed description for this PR:

@charliestock and I noticed that there were large biases in the surface silica values in ESM4.5 runs, particularly in the Southern Ocean. This appears to be associated with increases in phytoplankton mortality (`mort_Sm`, `mort_Md`, `mort_Lg`, `mort_Di` increased to 0.5% from 0 in ESM4.5 settings). The Si associated with phytoplankton mortality gets routed to dissolved pools (SiO4) instead of sinking down at a slow rate with sinking phytoplankton. Consequently, there ends up being insufficient Si detritus generation, resulting in a build-up of silica in the surface oceans. 

To resolve this we added a bit more flexibility in how Si is routed to dissolved vs particulate forms. 
First, we implemented a Si loss associated with negative growth, to be consistent with how N, P, and Fe losses due to respiration exceeding growth is handled. 
Second, we added a scaling factor (`phi_sidiss_mort`) that determines the amount of silica test left over in medium and large phytoplankton as phytoplankton dies. To target an overall dissolution rate of 0.01 day-1 at 0 degrees C (consistent with detritus) we can set this value to be 0.67. However, to reproduce answers, here in this PR we have set `phi_sidiss_mort` to be 0. 
Lastly, we added flexibility to what percentage of Si ingestion by zooplankton is egested to detritus. By default, the values of `phi_det_si` for small, medium, and large zooplankton are the same as `phi_det`, but this parameter can also be increased to route more silicon ingestion to detritus. For ESM4.5 we are testing values 5-10% more than the default `phi_det` values.

jessluo commented:
> Jessica, can you write a sentence or two for the commit message associated with these changes? Then I will be able to merge it.

Commit message can be "Adding flexibility to the routing of phytoplankton silicon losses to enable increases in the production of Si detritus."

charliestock commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -1040,6 +1040,13 @@ subroutine user_add_params(param_file)
     call get_param(param_file, "generic_COBALT", "mort_Md", phyto(MEDIUM)%mort, &
                    "mortality (cell death) rate constant for medium phytoplankton @ 0 deg. C", &
                    units="day-1", default=0.0, scale=I_sperd)
+    ! Diatom silica exudation or loss due to mortality and basal respiration
```
Could we make the names here consistent with jdissloss_si_Lg, Md etc.  e.g., "fraction of medium diatom silica dissolved during respiration and mortality"

charliestock commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -1360,7 +1367,14 @@ subroutine user_add_params(param_file)
                    "fraction of P ingestion by large zooplankton to semi-labile dissolved organic phosphorus", &
                    units="none", default=0.3*(0.30-zoo(3)%phi_det))
-    !
+    ! Partitioning of silica detritus production from zooplankton is by default slightly higher than organic matter detritus
```
Isn't the default the same for now? though I suspect we will change this after ESM4.5 is completed.

charliestock commented on generic_tracers/generic_COBALT.F90:
```diff
@@ -4501,7 +4524,9 @@ subroutine generic_COBALT_update_from_source(tracer_list,Temp,Salt,rho_dzt,dzt,h
-            phyto(n)%jmortloss_sio2(i,j,k) = phyto(n)%jmortloss_n(i,j,k)*phyto(n)%q_si_2_n(i,j,k)
+            ! silica dissolution from phytoplankton mortality is also multiplied by a scaling factor that 
+            ! determines the amount of silica test left over as the phytoplankton dies
+            phyto(n)%jdissloss_si(i,j,k) = phyto(n)%phi_sidiss_mort*phyto(n)%jmortloss_n(i,j,k)*phyto(n)%q_si_2_n(i,j,k)
```
Does this need to be added to phyto(n)%jdissloss_si(i,j,k) defined on line 3957 and 3958?  I think so, since the total dissolution from the phytoplankton is the sum of those two terms.

charliestock reviewed (changes requested):
Hello @jessluo and @theresa-cordero,  

I had a quick look at the latest version of this and I think there may be one issue: the phytoplankton dissolution from phytoplankton mortality on line 4529 must be added to the any dissolution resulting from net respiration on lines 3957 and 3958.  That is:

phyto(n)%jdissloss_si(i,j,k) =  phyto(n)%jdissloss_si(i,j,k) + phyto(n)%phi_sidiss_mort*phyto(n)%jmortloss_n(i,j,k)*phyto(n)%q_si_2_n(i,j,k)

On line 4529.  This wouldn't have tripped off any mass balance checks and makes no difference with the default parameters, but it would preserve a bit more silica from sinking phytoplankton than intended in ESM4.5.

I left one or two other minor comments, but I think this is almost there!

thank you,
Charlie

jessluo committed:
Fixes to the accounting of jdissloss_si

jessluo commented:
Fixed according to @charliestock's suggestions

charliestock reviewed (approved):
Thanks @jessluo, the changes look good to me.  Moving ahead with approval.
